shell/platform/embedder/embedder.cc

```diff
@@ -1583,3 +1584,116 @@ FlutterEngineResult FlutterEngineUpdateLocales(FLUTTER_API_SYMBOL(FlutterEngine)
+      break;
+    case kFlutterEngineDartObjectTypeInt64:
+      dart_object.type = Dart_CObject_kInt64;
+      dart_object.value.as_int64 = object->int64_value;
+      break;
+    case kFlutterEngineDartObjectTypeDouble:
+      dart_object.type = Dart_CObject_kDouble;
+      dart_object.value.as_double = object->double_value;
+      break;
+    case kFlutterEngineDartObjectTypeString:
+      if (object->string_value == nullptr) {
+        return LOG_EMBEDDER_ERROR(kInvalidArguments,
+                                  "kFlutterEngineDartObjectTypeString must be "
+                                  "a null terminated string but was null.");
+      }
+      dart_object.type = Dart_CObject_kString;
+      dart_object.value.as_string = const_cast<char*>(object->string_value);
+      break;
+    case kFlutterEngineDartObjectTypeBuffer: {
+      const auto* buffer = SAFE_ACCESS(object->buffer_value, buffer, nullptr);
+      if (buffer == nullptr) {
+        return LOG_EMBEDDER_ERROR(kInvalidArguments,
+                                  "kFlutterEngineDartObjectTypeBuffer must "
+                                  "specify a buffer but found nullptr.");
+      }
+      auto buffer_size = SAFE_ACCESS(object->buffer_value, buffer_size, 0);
+      auto callback =
+          SAFE_ACCESS(object->buffer_value, buffer_collect_callback, nullptr);
+      auto user_data = SAFE_ACCESS(object->buffer_value, user_data, nullptr);
+
+      // The the user has provided a callback, let them manage the lifecycle of
+      // the underlying data. If not, copy it out from the provided buffer.
+
+      if (callback == nullptr) {
+        dart_object.type = Dart_CObject_kTypedData;
+        dart_object.value.as_typed_data.type = Dart_TypedData_kUint8;
+        dart_object.value.as_typed_data.length = buffer_size;
+        dart_object.value.as_typed_data.values = const_cast<uint8_t*>(buffer);
+      } else {
+        struct ExternalTypedDataPeer {
+          void* user_data = nullptr;
+          VoidCallback trampoline = nullptr;
+        };
+        auto peer = new ExternalTypedDataPeer();
+        peer->user_data = user_data;
+        peer->trampoline = callback;
+        dart_object.type = Dart_CObject_kExternalTypedData;
+        dart_object.value.as_external_typed_data.type = Dart_TypedData_kUint8;
+        dart_object.value.as_external_typed_data.length = buffer_size;
+        dart_object.value.as_external_typed_data.data =
+            const_cast<uint8_t*>(buffer);
+        dart_object.value.as_external_typed_data.peer = peer;
+        dart_object.value.as_external_typed_data.callback =
+            +[](void* unused_isolate_callback_data,
+                Dart_WeakPersistentHandle unused_handle, void* peer) {
+              auto typed_peer = reinterpret_cast<ExternalTypedDataPeer*>(peer);
+              typed_peer->trampoline(typed_peer->user_data);
+              // TODO(chinmaygarde): Check if the finalizer gets called by the
```

Comment:
@rmacnak-google Can you clarify the behavior in the error case please? I'd need to manually invoke the embedder finalizer if the post fails below.

Reply:
Yes. If Dart_PostCObject returns success, the message was queued, and any finalizers in the message will eventually run, even if the receiving isolate shuts down before handling the message. If it returns failure, the external memory still belongs to the caller.

Reply:
Thanks. I clarified the docs and made sure to collect the embedder created peer.